network/dag/tree/iblt.go

```diff
@@ -0,0 +1,295 @@
+	}
+
+	// valid
+	return o, nil
+}
+
+// Decode tries to deconstruct the iblt into hashes remaining (positive entries) and missing (negative entries) from the iblt.
+// Decode is destructive to the iblt. If decoding fails with ErrDecodeNotPossible, the original iblt can be recovered by
+// Insert-ing all remaining and Delete-ing all missing hashes. Any other error is unrecoverable.
+func (i *Iblt) Decode() (remaining []hash.SHA256Hash, missing []hash.SHA256Hash, err error) {
+	for {
+		updated := false
+
+		// peel off pures (count == ±1)
+		for _, b := range i.Buckets {
+			if (b.Count == 1 || b.Count == -1) && i.hashKey(b.KeySum) == b.HashSum {
+				if b.Count == 1 {
+					remaining = append(remaining, b.KeySum)
+					err = i.Delete(b.KeySum)
+				} else { // b.count == -1
+					missing = append(missing, b.KeySum)
+					err = i.Insert(b.KeySum)
+				}
+				if err != nil {
+					return nil, nil, err
+				}
+				updated = true
+			}
+		}
+
+		// if no pures exist, the iblt is empty or cannot be decoded
+		if !updated {
+			for _, b := range i.Buckets {
+				if !b.isEmpty() {
+					return remaining, missing, ErrDecodeNotPossible
+				}
+			}
+			return remaining, missing, nil
+		}
+	}
+}
+
+func (i Iblt) numBuckets() int {
+	return len(i.Buckets)
+}
+
+func makeBuckets(numBuckets int) []*bucket {
+	buckets := make([]*bucket, numBuckets)
+	for i := 0; i < numBuckets; i++ {
+		buckets[i] = new(bucket)
+	}
+	return buckets
+}
+
+func (i Iblt) bucketIndices(hash uint64) []uint32 {
+	bucketUsed := make(map[uint32]bool, i.K)
+	var indices []uint32
+	hashKeyBytes, nextBytes := make([]byte, 8), make([]byte, 4)
+	binary.LittleEndian.PutUint64(hashKeyBytes, hash)
+	next := murmur3.SeedSum32(i.Hk, hashKeyBytes)
+	for len(indices) < int(i.K) {
+		bucketId := next % uint32(i.numBuckets())
+		if !bucketUsed[bucketId] {
+			indices = append(indices, bucketId)
+			bucketUsed[bucketId] = true
+		}
+		binary.LittleEndian.PutUint32(nextBytes, next)
+		next = murmur3.SeedSum32(i.Hk, nextBytes)
+	}
+	return indices
+}
+
+func (i Iblt) hashKey(key hash.SHA256Hash) uint64 {
+	return murmur3.SeedSum64(i.Hc, key.Slice())
+}
+
+func (i Iblt) MarshalBinary() ([]byte, error) {
+	data := make([]byte, i.numBuckets()*bucketBytes)
+	for idx, b := range i.Buckets {
+		bs, err := b.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		copy(data[idx*bucketBytes:], bs)
+	}
+	return data, nil
+}
+
+func (i *Iblt) UnmarshalBinary(data []byte) error {
+	numBuckets := len(data) / bucketBytes
+	if len(data) != numBuckets*bucketBytes {
+		return errors.New("invalid data length")
+	}
+	buf := bytes.NewBuffer(data)
+	i.Hc = ibltHc
+	i.Hk = ibltHk
+	i.K = ibltK
+	i.Buckets = makeBuckets(numBuckets)
+	for j := 0; j < i.numBuckets(); j++ {
+		err := i.Buckets[j].UnmarshalBinary(buf.Next(bucketBytes))
+		if err != nil {
+			return fmt.Errorf("unmarshalling failed - %w", err)
+		}
+	}
+	return nil
+}
+
+// bucket
+type bucket struct {
```

Comment:
json?

Reply:
initial marshaling was json. forgot to remove these tags